### PR TITLE
Introduce CacheKeyGenerator

### LINKED ISSUE #23515: Quarkus cache CompositeCacheKey only consists of method arguments - should include method name as well
### Description

Currently, the CompositeCacheKey of the Quarkus cache (caffeine) extension is calculating a deep hash on the list of method arguments. This is not sufficient if there are multiple different methods with the same signature.

Sample :  method getXYZ(String x, String z) and method getABC(String x, String z) would result in conflicting/equal CompositeCacheKeys leading to unexpected behaviour (e.g. wrong results are returned).

### Implementation ideas

Incl. the current method name as part of the CompositeCacheKey instead of relying on different argument names,

### LINKED ISSUE #25588: Use a particular field of a Class as CacheKey
### Description

I have a method having an Class as parameter. The documentation says the following
- If the method has no arguments, then the cache key is an instance of io.quarkus.cache.DefaultCacheKey built from the cache name.
- If the method has exactly one argument, then this argument is the cache key.
- If the method has multiple arguments but only one annotated with @CacheKey, then this annotated argument is the cache key.
- In all other cases, the cache key is an instance of io.quarkus.cache.CompositeCacheKey built from multiple method arguments (annotated with @CacheKey or not).

I would like to be able to use just one single field of the class as a CacheKey

### Implementation ideas

If a method is annotated with @CacheName, I could be able to have an annotation @CacheKey("name_of_parameter.name_of_field_in_class") where the parameter of the method is named "name_of_parameter" of type Clazz and it has a field  "name_of_field_in_class" that will be used as cache key or in the Class of name_of_parameter Clazz, I could place the annotation @CacheKey on the field name_of_field_in_class so that it can automatically by used as a cache key when passed to a method annotated with @CacheName.

That would be helpful.

### PR DESCRIPTION
**Pull request by gwenneg**
Fixes #23515 
Fixes #25588

See the [doc update](https://github.com/quarkusio/quarkus/blob/8bb2f9a5741f4806d07f4c1c7b0ad0ce3a3af68e/docs/src/main/asciidoc/cache.adoc#generating-a-cache-key-with-cachekeygenerator) for details about `CacheKeyGenerator`.

### CODE REVIEW

File: docs/src/main/asciidoc/cache.adoc [hunk 1]
```diff
@@ -323,7 +320,22 @@ method annotated with `@CacheResult` or `@CacheInvalidate`.
...
+=== Cache keys building logic
+
+Cache keys are built by the annotations API using the following logic:
+
+* If an `io.quarkus.cache.CacheKeyGenerator` is declared in a `@CacheResult` or a `@CacheInvalidate` annotation, then it is used to generate the cache key. The `@CacheKey` annotations that might be present on some of the method arguments are ignored.
+* Otherwise, if the method has no arguments, then the cache key is an instance of `io.quarkus.cache.DefaultCacheKey` built from the cache name.
+* Otherwise, if the method has exactly one argument, then that argument is the cache key.
+* Otherwise, if the method has multiple arguments but only one annotated with `@CacheKey`, then that annotated argument is the cache key.
+* Otherwise, if the method has multiple arguments annotated with `@CacheKey`, then the cache key is an instance of `io.quarkus.cache.CompositeCacheKey` built from these annotated arguments.
+* Otherwise, the cache key is an instance of `io.quarkus.cache.CompositeCacheKey` built from all the method arguments.
```
Comment (geoand) [inline docs/src/main/asciidoc/cache.adoc:333]: It's a little hard to follow all these `otherwise` statements, so maybe the last one should mention the conditions it applies to?

File: docs/src/main/asciidoc/cache.adoc [hunk 2]
```diff
@@ -366,6 +378,112 @@ public class CachedService {
 <3> Calling this method WILL invalidate values cached by the `load` method because the key elements order is the same.
 <4> Calling this method WILL NOT invalidate values cached by the `load` method because the key elements order is different.
 
+=== Generating a cache key with `CacheKeyGenerator`
+
+You may want to include more than the arguments of a method into a cache key.
+This can be done by implementing the `io.quarkus.cache.CacheKeyGenerator` interface and declaring that implementation in the `keyGenerator` field of a `@CacheResult` or `@CacheInvalidate` annotation.
+
+If a CDI scope is declared on a key generator, then it will be injected as a CDI bean during the cache key computation.
+Otherwise, the key generator will be instantiated using its default constructor.
```
Comment (geoand) [inline docs/src/main/asciidoc/cache.adoc:387]: I think this sentence is redundant given the warning below

File: docs/src/main/asciidoc/cache.adoc [hunk 3]
```diff
@@ -366,6 +378,112 @@ public class CachedService {
...
+All CDI scopes supported by Quarkus can be used on a key generator.
+
+[WARNING]
+====
+All cache key generators must have a default no-arg constructor.
```
Comment (geoand) [inline docs/src/main/asciidoc/cache.adoc:392]: If it's a CDI bean, why is this a requirement?

**Review by geoand — COMMENTED** [review]
Very nice improvement!

I added some docs related comments. I also think that a review from @mkouba would be beneficial

File: docs/src/main/asciidoc/cache.adoc (hunk 3 above)
Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc:392]: Good question. I guess I should lift that requirement for CDI beans.

I wonder what's the best approach to do that at build time though. Looking for CDI scopes annotations on the generator class isn't enough since the generator could still be a bean produced by CDI even if it didn't have any scope declared on its class. Is there an easy way to determine from a `@BuildStep` if an instance is a bean managed by CDI?

Comment (geoand) [reply docs/src/main/asciidoc/cache.adoc:392]: I think there is, but lets see what @mkouba has to say :)

Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc:392]: Maybe `Arc.container().instance(keyGeneratorClass).isAvailable()` works at build time, I didn't test it. I'll check that right away.

File: docs/src/main/asciidoc/cache.adoc (hunk 1 above)
Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc:333]: Do you think starting with `Otherwise, if the method has multiple arguments and none of them are annotated with @CacheKey` would be better? From a code perspective, that last sentence is a simple `else` block.

File: docs/src/main/asciidoc/cache.adoc (hunk 3 above)
Comment (geoand) [reply docs/src/main/asciidoc/cache.adoc:392]: Have you looked at `BeanDiscoveryFinishedBuildItem`?

File: docs/src/main/asciidoc/cache.adoc (hunk 1 above)
Comment (geoand) [reply docs/src/main/asciidoc/cache.adoc:333]: > would be better?

I do.

> From a code perspective, that last sentence is a simple else block

Yeah I understand that, I am just saying that it's very hard to follow that logic in english writing :)

File: docs/src/main/asciidoc/cache.adoc (hunk 2 above)
Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc:387]: I think this sentence is more important than the warning. Should I remove the warning? The users will quickly understand what's wrong with the build time exception anyway.

Comment (geoand) [reply docs/src/main/asciidoc/cache.adoc:387]: Agreed

File: docs/src/main/asciidoc/cache.adoc (hunk 3 above)
Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc:392]: No, but that name sounds promising 😃

File: docs/src/main/asciidoc/cache.adoc (hunk 2 above)
Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc:387]: Warning removed.

File: docs/src/main/asciidoc/cache.adoc (hunk 3 above)
Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc]: I updated `CacheProcessor`. CDI beans are now excluded from the no-arg constructor check for key generators.

File: extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java [hunk 4]
```diff
@@ -102,6 +114,7 @@ void validateCacheAnnotationsAndProduceCacheNames(CombinedIndexBuildItem combine
         for (DotName bindingName : INTERCEPTOR_BINDINGS) {
             for (AnnotationInstance binding : combinedIndex.getIndex().getAnnotations(bindingName)) {
                 throwables.addAll(validateInterceptorBindingTarget(binding, binding.target()));
+                findCacheKeyGenerator(binding, binding.target()).ifPresent(keyGenerators::add);
```
Comment (gastaldi) [inline extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:117]: Better avoid lambdas and method references here

Comment (gwenneg) [reply extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:117]: Does the "no-lambda policy" apply to build time code now? Last time I heard about it, it was only for run time code.

Comment (geoand) [reply extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:117]: It's runtime only

Comment (gastaldi) [reply extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:117]: Yes, runtime. I haven't noticed that this is build time code

File: docs/src/main/asciidoc/cache.adoc [hunk 5]
```diff
@@ -366,6 +378,106 @@ public class CachedService {
 <3> Calling this method WILL invalidate values cached by the `load` method because the key elements order is the same.
 <4> Calling this method WILL NOT invalidate values cached by the `load` method because the key elements order is different.
 
+=== Generating a cache key with `CacheKeyGenerator`
+
+You may want to include more than the arguments of a method into a cache key.
+This can be done by implementing the `io.quarkus.cache.CacheKeyGenerator` interface and declaring that implementation in the `keyGenerator` field of a `@CacheResult` or `@CacheInvalidate` annotation.
+
+If a CDI scope is declared on a key generator, then it will be injected as a CDI bean during the cache key computation.
```
Comment (mkouba) [inline docs/src/main/asciidoc/cache.adoc:386]: So I suppose that only class-based [managed beans](https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html#managed_beans) are supported, right? I.e. no producers or synthetic beans.

Also it seems that you don't take CDI qualifiers into account at runtime and so you should probably forbid the usage of custom qualifiers, or just mention that the _default qualifier_ must be used (in CDI this means either no qualifier annotation or `@javax.enterprise.inject.Default`).

File: extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java [hunk 6]
```diff
@@ -186,6 +205,33 @@ private List<Throwable> validateInterceptorBindingTarget(AnnotationInstance bind
...
+                return Optional.of(keyGenerator.asClass().name());
+            }
+        }
+        return Optional.empty();
+    }
+
+    // Key generators must have a default constructor if they are not managed by Arc.
+    private List<Throwable> validateKeyGeneratorsDefaultConstructor(CombinedIndexBuildItem combinedIndex,
+            BeanDiscoveryFinishedBuildItem beanDiscoveryFinished, Set<DotName> keyGenerators) {
+        List<DotName> managedBeans = beanDiscoveryFinished.geBeans()
```
Comment (mkouba) [inline extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:221]: If only managed beans are supported then you should filter out unsupported bean types, i.e. `beanDiscoveryFinished.geBeans().stream().filter(BeanInfo::isClassBean)`.

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java [hunk 7]
```diff
@@ -138,6 +146,30 @@ protected Object getCacheKey(Cache cache, List<Short> cacheKeyParameterPositions
         }
     }
 
+    private Object generateKey(Class<? extends CacheKeyGenerator> keyGeneratorClass, Method method,
+            Object[] methodParameterValues) {
+        InstanceHandle<? extends CacheKeyGenerator> keyGeneratorHandle = Arc.container().instance(keyGeneratorClass);
```
Comment (mkouba) [inline extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java:151]: You can `@Inject Instance<CacheKeyGenerator>` directly in the interceptor and then use `instance.select(keyGeneratorClass)` and `Instance#isResolvable()` and `Instance#get()` instead. This way you would not need to produce an `UnremovableBeanBuildItem` for all the `CacheKeyGenerator` implementations and moreoever `@Dependent` scoped impls would be correctly bound the lifecycle of the interceptor.

Comment (gwenneg) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java:151]: I will still need to call `Instance#destroy` on the keygen instance after I'm done generating the cache key, right? I just tested your suggestion (which is great, thanks!) without calling `destroy` locally and the instances of `CacheKeyGenerator` are piling up in the memory.

File: docs/src/main/asciidoc/cache.adoc (hunk 5 above)
Comment (gwenneg) [reply docs/src/main/asciidoc/cache.adoc:386]: Yes, the key generators are restricted to class-based managed beans because we need to declare the keygen class in the caching annotations.

Qualifiers should be forbidden indeed and/or documented.

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java (hunk 7 above)
Comment (mkouba) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java]: Yes, for `@Dependent`  keygens you'd  need to call `destroy()` or even `@Inject io.quarkus.arc.InjectableInstance<CacheKeyGenerator>` and use `getHandle()` to obtain an `io.quarkus.arc.InstanceHandle` which also extends `AutoCloseable`.

**Review by mkouba — APPROVED** [review]
Looks good! :-)

**Comment by gwenneg** [comment]
Thanks everyone for the reviews!
